### ISSUE TITLE
Batch Job Priority is ignored

### ISSUE THREAD

**Issue by Fuco1** [hashicorp/nomad#12792]
### Nomad version
Nomad v1.2.6 (a6c6b475db5073e33885377b4a5c733e1161020c)

### Operating system and Environment details

Linux ubuntu in Azure.

### Issue

With batch jobs, I would expect the allocations from higher priority job to be placed first.

In fact, I've just rewrote a bunch of our infra related to batch job submission only to find out this is not the case :man_facepalming:  Is there any way to make sure the batch job allocations with higher priority are placed first? 

We're running 100k+ task groups doing time series forecasting but I really need to be able to just run it and forget it and be sure that what goes in first comes out first, otherwise I run 10 jobs and instead of waiting 1 hours for the first to finish, I wait 10 hours for all of them to finish.

See also https://github.com/hashicorp/nomad/issues/9845

### Reproduction steps

Submit two batch jobs with different priorities.

#### Expected Result

Allocations from higher priority job are placed first.

#### Actual Result

Allocations are placed at random.

**Comment by lgfa29** [comment]
Hi @Fuco1 👋 

The `priority` value defines two things:

1. the order in which evaluations are dequeued for scheduling 
2. the order of importance of jobs when the cluster is in resource contention

The doc page you linked refer to `1`. The Nomad leader has an internal queue of evaluations that need to be processed in order to, among other things, create and schedule allocations. A job with a high priority will "skip" this queue and jump to the front to be processed before lower priority jobs.

What this means is that this evaluation will be _processed_ first, but that doesn't mean it will _start_ first. Different jobs, clients, allocations etc. have different characteristics, so the allocation will take different time to start.

If your cluster is in a relatively healthy state, this queue will be pretty small, so the priority value won't have much effect. You can monitor the `nomad.nomad.plan.queue_depth` [metric](https://www.nomadproject.io/docs/operations/metrics-reference#key-metrics) to get a sense of what's going on in the queue.

For `2`, that's the what's described in the [comment in the issue you linked](https://github.com/hashicorp/nomad/issues/9845#issuecomment-762857009). When the Nomad cluster is low on resources and you try to schedule something that won't fit, Nomad will use the job priority to determine which allocations can be stopped to make room for the new ones. This is what's called [preemption](https://www.nomadproject.io/docs/internals/scheduling/preemption).

> We're running 100k+ task groups doing time series forecasting but I really need to be able to just run it and forget it and be sure that what goes in first comes out first, otherwise I run 10 jobs and instead of waiting 1 hours for the first to finish, I wait 10 hours for all of them to finish.

I didn't quite follow this part. Are you looking to have the 10 jobs run one after the other in a predefined sequence?

**Comment by Fuco1** [comment]
> I didn't quite follow this part. Are you looking to have the 10 jobs run one after the other in a predefined sequence?

I basically want a FIFO, which I now tried to achieve by submitting first job with priority 100, second job with priority 99 and so on, decreasing priority by 1 with each new job.  This limits us to 100 jobs but that is fine as we can further "scale" this with different constrains on the VMs for example.

Here is the concrete situation.

1. We spin up VMs worth 1000 CPUs and we schedule one task group per CPU.
2. Two users submit a new forecast job.  The first job has 10k task groups and priority 100.  Second job has 10k task groups and priority 99.
3. Nomad will start allocating the task groups from either job seemingly randomly.  After 1000 allocations are made (to exhaust the 1000 CPUs), the rest is placed as soon as something finishes.

This is all fine, except that the first user want to be done with the task *completely* before the second user.  Now I think if I simply make the priority 10 apart, the allocations from job 2 should be killed and preempted.  In fact I'm testing this right now, will let you know.

If that is true, it would be nice if we could somehow make nomad's "10 apart" configurable, because for our use-case, we run this cluster only for batch jobs and everything is opaque to the user, so they will never encounter any priorities or whatever, we generate that in the background.  They just submit some "work" and we generate the corresponding nomad jobs.

**Comment by Fuco1** [comment]
I tried it with the priorities being 10 apart and the lower priority allocations were preempted, but then never resumed computation even after the higher priority job completely finished (which is not technically true because it had 4 allocations stuck at starting for over 30 minutes, at which point I just killed it).

Overall the scheduling was very chaotic as Nomad was constantly placing the lower priority allocations only to kill them moments later when a higher priority task finished and tried to schedule from its job's queue.

I think Nomad should not even attempt placement if there are queued tasks from higher priority batch job.  This is what I expected to be the case.

**Comment by nh2** [comment]
I've made PR #13835 to better document the current behaviour, but would find it fantastic if this issue could be used to track actually implementing starting Batch jobs in priority order.

So far I had to build an awkward system of tags and [`constraint`](https://www.nomadproject.io/docs/job-specification/constraint)s to separate low-priority batch jobs from higher priority ones.

**Comment by cinnayyannic** [comment]
Hi, i guess batch job priority has not changed after the last comment ?
If not i just try to understand a bit better how batch scheduling works in general.

Lets assume i have an exhausted cluster and dispatch n new job runs.
Since the priority field has no effect on batch jobs how is the placement order then determined ?

From the comments above it does not seem to be strictly fifo.

**Comment by tgross** [comment]
> Hi, i guess batch job priority has not changed after the last comment ?

Correct, this work has not yet been done. It's a surprising large project because it'd require redesign of the eval broker.

> If not i just try to understand a bit better how batch scheduling works in general.

Each time a job is registered a new evaluation is created and enqueued in the eval broker. Scheduler workers on all the Nomad servers _concurrently_ dequeue evaluations and then submit "plans" to the leader to be applied serially. Once a plan is applied the clients pull down the allocations they need and start them up. There are 3 controls over concurrency:

* There are a limited number of scheduler workers (one per core on each follower and 25% of the that on the leader)
* The broker ensures that at most one eval for a given job is in-flight at a time (note that each dispatch of a job is its own job)
* The broker hands out higher-priority evals first.

Because the scheduler work is being done concurrently, there's no guarantee that a higher priority job will be _placed_ before a lower priority job, only that Nomad _starts_ scheduling that workload earlier.

You can also set the scheduler config (ex. with [`operator scheduler set-config -preempt-batch-scheduler`](https://developer.hashicorp.com/nomad/commands/operator/scheduler/set-config#preempt-batch-scheduler) but that allows higher-priority jobs to preempt allocations of lower-priority jobs that have already been placed. That is, it kicks out the lower-priority job.

So, moving to your example...

> Lets assume i have an exhausted cluster and dispatch n new job runs.
> Since the priority field has no effect on batch jobs how is the placement order then determined ?
> 
> From the comments above it does not seem to be strictly fifo.

That exact order of placement depends on a number of factors:
* Are scheduler workers idle? That is, is the rate of dispatch faster than the rate at which scheduler workers can complete the scheduling work and plans can be applied? If a scheduler worker is available the total window of time between "the job is registered" and "the client receives the alloc" is short enough that it's entirely possible, if you're dispatch rate is low enough, to see FIFO behavior.
* If not, then higher-priority jobs will get dequeued first. That means the scheduler workers will start chewing on them first, but not that they will get placed first. That order is not guaranteed.
* Plans are applied serially and at that point the resources for the allocations have been reserved, but that doesn't mean the allocations will actually spin-up in that order, because that's happening in parallel across all nodes.

If you want to guarantee FIFO order, for now the only workaround is an upstack abstraction where you dispatch and wait for the result of that eval. The "advantage" of this is that you can wait on arbitrary events, like for allocations to be healthy or even complete, and at that point you've effectively built yourself the long-desired https://github.com/hashicorp/nomad/issues/545. The disadvantage is of course that you'd be building it yourself. ☹️ This is something that is on our radar but hasn't managed to make its way up to the top of the priority stack for development.

**Comment by cinnayyannic** [comment]
Thanks a lot for the detailed explanation :)

I initially thought that for batch jobs the scheduler queues will fill up in case of exhausted clients, then priority could still have been used to sort the queue.

`If you want to guarantee FIFO order, for now the only workaround is an upstack abstraction where you dispatch and wait for the result of that eval. The "advantage" of this is that you can wait on arbitrary events, like for allocations to be healthy or even complete, and at that point you've effectively built yourself the long-desired https://github.com/hashicorp/nomad/issues/545.`

This is exactly what i am currently working on. I have batch tasks which have to be executed in a dependency order.
Lets say i have a execution graph with the Jobs A -> B ->C
Implementing the order is not much of a problem because Job A can be registered, custom scheduler implementation waits until Job A is finished and registers the next one.

What i was hoping to somehow handle with the job priority was weighting job graphs.

Lets say i have the following job graphs and the clients are already exhausted with previously triggered runs:

Graph 1)
A (prio 1) -> B (prio 2) -> C (prio 3) -> D (prio 4)
  
Graph 2)
E (prio 1) -> F (prio 2) -> G (prio 3) -> H (prio 4)

The start of a graph execution is triggered by the users. Graph 1) is started and after a while it reaches C, while C is running a user starts Graph 2 and the custom scheduler will register Job E (cluster still exhausted so E ist not running). Now Job C is finished and custom scheduler registers D.

With priority working the scheduling would favor placement of D (try to finish started work) instead of starting with the second graph.

**Comment by schmichael** [comment]
FYI: This issue is being discussed as part of #27927
